R/load_data.R

```diff
@@ -188,54 +208,53 @@ ReadMPX_Seurat <- function (
       cg_assay@colocalization <- colocalization
     }
   } else {
-    cg_assay <- CreateAssayObject(counts = X)
+    cg_assay <- switch(
+      getOption("Seurat.object.assay.version", "v3"),
+      "v3" = CreateAssayObject(counts = X),
+      "v5" = CreateAssay5Object(counts = X)
+    )
     Key(cg_assay) <- paste0(assay, "_")
   }
 
   # Create Seurat object
-  seur_obj <- CreateSeuratObject(counts = cg_assay, assay = assay, ...)
+  seur_obj <- CreateSeuratObject(counts = cg_assay, assay = assay)#, ...)
   if (verbose && check_global_verbosity())
-    cli_alert_success("Created a 'Seurat' object with {col_br_blue(ncol(X))} cells and {col_br_blue(nrow(X))} targeted surface proteins")
-
-  if (add_additional_assays) {
-    for (layr in names(anndata_hier$obsm)) {
-
-      # Add only normalized counts as individual assays
-      if (!layr %in% c("denoised", "normalized_clr", "normalized_rel")) next
-
-      X_assay <-
-        anndata_hier$obsm[[layr]] %>%
-        as_tibble() %>%
-        column_to_rownames("component") %>%
-        t()
-      layer_assay <- CreateAssayObject(counts = X_assay)
-      suppressWarnings(seur_obj[[layr]] <- layer_assay)
-    }
-  }
-
+    cli_alert_success(glue("Created a 'Seurat' object with {col_br_blue(ncol(X))} cells ",
+                           "and {col_br_blue(nrow(X))} targeted surface proteins"))
 
   # Extract meta data
   seur_obj@meta.data <-
-    anndata_hier$obs %>%
-    {
-      newdata <- .
-      for (name in names(newdata)) {
-        if (length(newdata[[name]]) == dim(anndata_hier$X)[2]) next
-        indicies <- newdata[[name]]$codes + 1
-        categories <- newdata[[name]]$categories
-        newdata[[name]] <- factor(categories[indicies], categories)
+    names(hd5_object[["obs"]]) %>% lapply(function(nm) {
+      if (length(names(hd5_object[["obs"]][[nm]])) == 2) {
+        indices <- hd5_object[["obs"]][[nm]][["codes"]]$read() + 1
+        categories <- hd5_object[["obs"]][[nm]][["categories"]]$read()
+        col <- tibble(!!sym(nm) := factor(categories[indices], categories))
```

Comment:
What is going on here?

Reply:
Hehe I'll try to add some comments in the code.

I have replaced rhdf5 with hdf5r here. It has a pretty strange API but I think it will be useful for us to get rid of rhdf5 since it doesn't play well with arm chip sets. 

But what the code does is to go through the "obs" table in the anndata file and convert them into tibble columns. This part:

````
indices <- hd5_object[["obs"]][[nm]][["codes"]]$read() + 1
        categories <- hd5_object[["obs"]][[nm]][["categories"]]$read()
````

converts dictionaries into factors. In the anndata object, the dictionaries have a codes vector (values) and a categories vector (keys).  If an element in "obs" has length 2, it's a dictionary that should be converted into a factor.

This stuff `!!sym(nm) := ` is just a trick to create a name from a variable. For example, if `nm = "leiden"`, these two solutions should be equivalent:

````
tibble(leiden = factor(categories[indices], categories))
tibble(!!sym(nm) := = factor(categories[indices], categories))
````

Reply:
Check, thanks!